libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,275 @@
+	}
+
+	// Path compression updates the ancestors of vertices along
+	// the path to the ancestor with the minimum label value.
+	void compressPath(
+		std::vector<size_t> &_ancestor,
+		std::vector<size_t> &_label,
+		std::vector<size_t> &_semi,
+		size_t _v
+	)
+	{
+		solAssert(_ancestor[_v] != std::numeric_limits<size_t>::max());
+		size_t u = _ancestor[_v];
+		if (_ancestor[u] != std::numeric_limits<size_t>::max())
+		{
+			compressPath(_ancestor, _label, _semi, u);
+			if (_semi[_label[u]] < _semi[_label[_v]])
+				_label[_v] = _label[u];
+			_ancestor[_v] = _ancestor[u];
+		}
+	}
+
+	std::vector<size_t> lengauerTarjanDominator(Vertex const& _entry, size_t numVertices)
+	{
+		solAssert(numVertices > 0);
+		// semi(w): The dfs index of the semidominator of ``w``.
+		std::vector<size_t> semi(numVertices, std::numeric_limits<size_t>::max());
+		// parent(w): The index of the vertex which is the parent of ``w`` in the spanning
+		// tree generated by the dfs.
+		std::vector<size_t> parent(numVertices, std::numeric_limits<size_t>::max());
+		// ancestor(w): The highest ancestor of a vertex ``w`` in the dominator tree used
+		// for path compression.
+		std::vector<size_t> ancestor(numVertices, std::numeric_limits<size_t>::max());
+		// label(w): The index of the vertex ``w`` with the minimum semidominator in the path
+		// to its parent.
+		std::vector<size_t> label(numVertices, 0);
+
+		// ``link`` adds an edge to the virtual forest.
+		// It copies the parent of w to the ancestor array to limit the search path upwards.
+		// TODO: implement sophisticated link-eval algorithm as shown in pg 132
+		// See: https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+		auto link = [&](size_t _parent, size_t _w)
+		{
+			ancestor[_w] = _parent;
+		};
+
+		// ``eval`` computes the path compression.
+		// Finds ancestor with lowest semi-dominator dfs number (i.e. index).
+		auto eval = [&](size_t _v) -> size_t
+		{
+			if (ancestor[_v] != std::numeric_limits<size_t>::max())
+			{
+				compressPath(ancestor, label, semi, _v);
+				return label[_v];
+			}
+			return _v;
+		};
+
+		auto toIdx = [&](Vertex const& v) { return m_vertexIndex[v]; };
+
+		// step 1
+		std::set<Vertex> visited;
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<size_t>> predecessors(numVertices);
+		// bucket(w): a set of vertices whose semidominator is ``w``
+		// The index of the array represents the vertex's ``dfIdx``
+		std::vector<std::deque<size_t>> bucket(numVertices);
+		// idom(w): the index of the immediate dominator of ``w``
+		std::vector<size_t> idom(numVertices, std::numeric_limits<size_t>::max());
+		// The number of vertices reached during the dfs.
+		// The vertices are indexed based on this number.
+		size_t dfIdx = 0;
+		auto dfs = [&](Vertex const& _v, auto _dfs) -> void {
+			if (visited.count(_v))
+				return;
+			visited.insert(_v);
```

Comment:
Hahaha right, I will change to the first suggestion since it is the same also used here: https://github.com/ethereum/solidity/blob/v0.8.21/libsolutil/Algorithms.h#L110 and I will make this code non-recursive and move it to `Algorithms.h` as suggested by Kamil.

Reply:
I think that the second version is clearer. IMO the code is more readable if actions with side effects are kept outside of `if`s.

Reply:
I only posted the first version as a joke, the second one is definitely the way to go :)

Reply:
Well, you could say that there a lot of funny places in our current codebase :)